tests/rptest/tests/pandaproxy_test.py

```diff
@@ -2129,3 +2136,243 @@ def test_upgrade_and_enable_basic_auth(self, base_release: tuple[int, int],
+        self.auth_tuple = auth_tuple
+        self.result_raw = None
+        self.partition_offsets = {}
+        self.done = False
+        self.exception = None
+
+    def run(self):
+        def do_produce():
+            msg_ascii = self.topic._random_topic_suffix(
+                size=self.bytes_per_second)
+            data = {"records": [{"value": msg_ascii}]}
+            self.logger.debug(f"Producing {json.dumps(data)}")
+            self.result_raw = self._produce_topic(
+                self.topic.name,
+                json.dumps(data),
+                headers=HTTP_PRODUCE_JSON_V2_TOPIC_HEADERS,
+                auth=self.auth_tuple)
+            assert self.result_raw.status_code == requests.codes.ok
+            produce_result = self.result_raw.json()
+            assert "offsets" in produce_result
+            assert len(produce_result["offsets"]) == 1
+            offset = produce_result["offsets"][0]
+            assert "error_code" not in offset, offset["error_code"]
+            if offset["partition"] not in self.partition_offsets:
+                self.partition_offsets[offset["partition"]] = {}
+            self.partition_offsets[offset["partition"]][
+                offset["offset"]] = msg_ascii
+            self.logger.debug(
+                f"Producer {self.name}: added {json.dumps(produce_result)}")
+
+            return False
+
+        try:
+            wait_until(do_produce, timeout_sec=self.timeout_sec, backoff_sec=1)
+        except ducktape.errors.TimeoutError as ex:
+            self.done = True
+        except Exception as ex:
+            self.exception = ex
+
+
+class VerifiablePandaproxyConsumer(threading.Thread):
+    def __init__(self,
+                 cons_name: str,
+                 topic: TopicSpec,
+                 consumer: Consumer,
+                 verifiable_producers: list[VerifiablePandaproxyProducer],
+                 timeout_sec: int,
+                 logger: TestLoggerMaker,
+                 auth_tuple: Optional[tuple[str, str]] = None):
+        threading.Thread.__init__(self, name=cons_name)
+        self.daemon = True
+        self.topic = topic
+        self.consumer = consumer
+        self.verifiable_producers = verifiable_producers
+        self.timeout_sec = timeout_sec
+        self.logger = logger
+        self.auth_tuple = auth_tuple
+        self.result_raw = None
+        self.done = False
+        self.exception = None
+
+    def _filter_producer_by_partition_offset(self, partition: int,
+                                             offset: int):
+        try:
+            prod = next(
+                filter(
+                    lambda prod: partition in prod.partition_offsets and offset
+                    in prod.partition_offsets[partition],
+                    self.verifiable_producers))
+            return prod
+        except StopIteration:
+            raise RuntimeError(
+                f"Failed to find producer with partition-offset {partition}-{offset}"
```

Comment:
A lot of CI failures here.

Reply:
This is happening because of a race to appending the partition-offsets into the producers offsets list. To be clear, here are the steps.

In producer:
1. Send a record
2. Verify the result is OK
3. Append the partition-offset to list

In consumer:
1. Fetch records
2. Verify the result is OK
3. Verify the fetched partition-offset is in the producers list

It looks like the consumer may verify the fetched partition-offset (consumer step 3) before python has the chance to append the partition-offset to the producers list (producer step 3).

The solution is simple, we can either:
A. add a mutex to the producer list and call that mutex in the consumer
B. retry consumer step 3 with a timeout

I will go with solution (B) because I believe (A) does not scale well when we have more than one producer and consumer